arpeggiator sorted mode fixed

szeli1 opened the pull request:
closes #6499
closes #4491

with this change the arpeggiator with sorted mode enabled will cycle through all of the active (currently playing) note's chords in order without skipping notes or randomly playing other ones.

this works by combining all of the chord's keys in an array and then playing it back as one single larger chord (sorted as suggested).

zonkmachine commented:
Hi! Testing this now (sorry for the long waiting time). Thank you so much for taking care of this issue! It seem to work really well but the arpeggiator has a lot of functions now so it's starting to get really finicky to test. While testing this PR I discovered a bug that I introduced with the 'Repeats' function. This has been fixed in https://github.com/LMMS/lmms/pull/7007 together with the original bug which that PR takes care of. I think it's better to merge that PR first and then rebase this PR so we can test this one some more with two less bugs to worry about. Are you OK with this?

zonkmachine commented:
I'm currently running these changes on top from https://github.com/LMMS/lmms/pull/7007

```cpp
diff --git a/src/core/InstrumentFunctions.cpp b/src/core/InstrumentFunctions.cpp
index 0911d28a4..b1fa0902e 100644
--- a/src/core/InstrumentFunctions.cpp
+++ b/src/core/InstrumentFunctions.cpp
@@ -470,42 +470,25 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
 
                int cur_arp_idx = 0;
                // process according to arpeggio-direction...
-               if( dir == ArpDirection::Up )
+               if (dir == ArpDirection::Up || dir == ArpDirection::Down)
                {
                        cur_arp_idx = (cur_frame / arp_frames) % (range + sortOffset);
                }
-               else if( dir == ArpDirection::Down )
-               {
-                       cur_arp_idx = range + sortOffset - (cur_frame / arp_frames) %
-                                                               (range + sortOffset) - 1;
-               }
-               else if (dir == ArpDirection::UpAndDown && (range + sortOffset) > 1)
+               else if ((dir == ArpDirection::UpAndDown || dir == ArpDirection::DownAndUp) && (range + sortOffset) > 1)
                {
                        // imagine, we had to play the arp once up and then
                        // once down -> makes 2 * range possible notes...
                        // because we don't play the lower and upper notes
                        // twice, we have to subtract 2
-                       cur_arp_idx = (cur_frame / arp_frames) % ((range + sortOffset) * 2 - 2);
+                       cur_arp_idx = (cur_frame / arp_frames) % ((range + sortOffset) * 2 - (2 * static_cast<int>(m_arpRepeatsModel.value())));
                        // if greater than range, we have to play down...
                        // looks like the code for arp_dir==DOWN... :)
                        if (cur_arp_idx >= range + sortOffset)
                        {
                                cur_arp_idx = range + sortOffset - cur_arp_idx % (range + sortOffset - 1) - 1;
+                               cur_arp_idx = range + sortOffset - cur_arp_idx % (range + sortOffset - 1) - static_cast<int>(m_arpRepeatsModel.value());
                        }
                }
-               else if (dir == ArpDirection::DownAndUp && (range + sortOffset) > 1)
-               {
-                       // copied from ArpDirection::UpAndDown above
-                       cur_arp_idx = (cur_frame / arp_frames) % ((range + sortOffset) * 2 - 2);
-                       // if greater than range, we have to play down...
-                       // looks like the code for arp_dir==DOWN... :)
-                       if (cur_arp_idx >= range + sortOffset)
-                       {
-                               cur_arp_idx = range + sortOffset - cur_arp_idx % (range + sortOffset - 1) - 1;
-                       }
-                       // inverts direction
-                       cur_arp_idx = range + sortOffset - cur_arp_idx - 1;
-               }
                else if( dir == ArpDirection::Random )
                {
                        // just pick a random chord-index
@@ -522,6 +505,12 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
                        cur_arp_idx %= static_cast<int>((range + sortOffset) / m_arpRepeatsModel.value());
                }
 
+               // If ArpDirection::Down or ArpDirection::DownAndUp, invert the final range.
+               if (dir == ArpDirection::Down || dir == ArpDirection::DownAndUp)
+               {
+                       cur_arp_idx = static_cast<int>(range + sortOffset / m_arpRepeatsModel.value()) - cur_arp_idx - 1;
+               }
+
                // now calculate final key for our arp-note
                int sub_note_key = 0;
                if (static_cast<ArpMode>(m_arpModeModel.value()) != ArpMode::Sort)
(END)

```

szeli1 commented:
> Are you OK with this?

What does this mean for me? If it isn't too difficult, then sure! (I do not have this branch locally currently)

zonkmachine commented:
I suspect it's a simple rebase. I can push the changes to your branch if you like. 
Can you rebase this PR on latest master?

szeli1 commented:
Do what you think is the best solution here. I will try to rebase soon.

zonkmachine commented:
OK - The plan is:
1. you rebase this PR
2. We merge https://github.com/LMMS/lmms/pull/7007
3. This PR is rebased once more on master

szeli1 commented:
I think i have rebased. You can merge #7007 if I'm correct.

zonkmachine commented:
> I think i have rebased. You can merge #7007 if I'm correct.

That looks correct. I'll allow some time for feedback on that PR first. A day or two.

zonkmachine commented:
> This PR is rebased once more on master

I went with merging just my changes from https://github.com/LMMS/lmms/pull/7007. This keeps your original commits so we can more easily test for regressions.

zonkmachine commented on src/core/InstrumentFunctions.cpp:
```diff
@@ -474,37 +474,19 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
-			cur_arp_idx = (cur_frame / arp_frames) % ((range + sortOffset) * 2 - 2);
+			cur_arp_idx = (cur_frame / arp_frames) % ((range + sortOffset) * 2 - (2 * static_cast<int>(m_arpRepeatsModel.value())));
 			// if greater than range, we have to play down...
 			// looks like the code for arp_dir==DOWN... :)
```
// looks like the code for arp_dir==DOWN... :)

Which was removed so this line should be removed.

szeli1 replied:
I will fix this in a few days. Sorry for the waiting.

zonkmachine commented:
It looks like this could be a fix also for issue https://github.com/LMMS/lmms/issues/2606. At least the demo project in the original post is fixed.

szeli1 commented:
I have rebased, and changed your changes. This should be tested because unfortunately I do not remember my changes in great detail. There seems to be a display bug when sorted mode is enabled and multiple notes are pressed. This bug makes the pressed notes light up incorrectly (while not being played).

zonkmachine commented on src/core/InstrumentFunctions.cpp:
```diff
@@ -435,26 +472,27 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
 		if (dir == ArpDirection::Up || dir == ArpDirection::Down)
 		{
-			cur_arp_idx = ( cur_frame / arp_frames ) % range;
+			cur_arp_idx = (cur_frame / arp_frames) % (range + sortOffset);
```
`range + sortOffset` is the same as `total_range`. I think you can skip `sortOffset` alltogether.

szeli1 replied:
It is only total_range when sorting.

zonkmachine replied:
Ah, got it!

zonkmachine commented:
In this test project from issue https://github.com/LMMS/lmms/issues/2606 you can see the problems with Sync mode over more than one note. It has two identical arpeggios but one inverted so the sound cancels out unless one of the tracks skips a note. The last 6th of the sequence is the one with Sync mode over two notes. Turn up the tempo and enjoy!
[arpnotesilence-01.mmp.zip](https://github.com/LMMS/lmms/files/5682517/arpnotesilence-01.mmp.zip)

szeli1 commented:
What are you trying to say? Would you rather me fix that issue as well? And the display issue I talked about? If not, I think this is ready for merge.

zonkmachine commented:
> What are you trying to say? Would you rather me fix that issue as well?

No, you're welcome to, but I mentioned above that it looks like at least one of the demo projects was fixed by this commit so there is a connection. I'm a bit tired now so I'm happy to stick with just this issue.

I've tested it with a bunch of arpeggios and they are almost all possible to compute ahead what the result will be. Almost all. This doesn't seem to be true for when you have a note that is looped before it's end so that it will overlap itself. I don't think it's reasonable to cover everything and this is an edge case. Just seeing the same result over and over again is fine.
Example: [arpdowncycle-02.mmp.txt](https://github.com/LMMS/lmms/files/14968717/arpdowncycle-02.mmp.txt)

zonkmachine commented:
Fixes https://github.com/LMMS/lmms/issues/4491

Veratil commented on src/core/InstrumentFunctions.cpp:
```diff
@@ -395,15 +399,48 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
+			}
+
+			// sorting:
+			std::sort(noteKeysArray.begin(), noteKeysArray.end(), [](int a, int b)
+			{
+				return a < b;
+			});
```
Is it possible to just use `std::less`?
```suggestion
			std::sort(noteKeysArray.begin(), noteKeysArray.end(), std::less{});
```

szeli1 replied:
Replaced lambda with std::less.

Veratil commented on src/core/InstrumentFunctions.cpp:
```diff
@@ -464,18 +499,33 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
+		else
+		{
+			// if we are sorting, we already have the base note key and the chord key
+			sub_note_key = noteKeysArray[cur_arp_idx % total_chord_size] + (cur_arp_idx / total_chord_size) *
```
Compiler is probably smart and will get the quotient and remainder at the same time, but just in case we may want to use `std::div` and for `cur_arp_idx` and `total_chord_size`? Similar for above block with `cur_chord_size` denominator.

szeli1 replied:
I think I do not fully understand this. Could you please describe it in an other way?

Veratil replied:
Using `std::div` we pull the `%` and `/` operations out. Can make it easier to read. This isn't always needed since the operations are nearby and the compiler should be able to automatically do this for us, but it doesn't hurt readability and standardization.
```cpp
auto result = std::div(cur_arp_idx, total_chord_size);
sub_note_key = noteKeysArray[result.rem] + result.quot * ...;
```
Alternatively it could be done just by pulling the operations out:
```cpp
auto noteIndex = cur_arp_idx % total_chord_size;
auto offset = cur_arp_idx / total_chord_size;
sub_note_key = noteKeysArray[noteIndex] + offset * ...;
```

szeli1 replied:
Thanks for the explanation. I have implemented this.

Veratil commented on src/core/InstrumentFunctions.cpp:
```diff
@@ -435,26 +469,27 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
 		{
 			// just pick a random chord-index
-			cur_arp_idx = (int)( range * ( (float) rand() / (float) RAND_MAX ) );
+			cur_arp_idx = (int)((range + sortOffset) * ((float) rand() / (float) RAND_MAX));
```
Convert to static_cast
```suggestion
			cur_arp_idx = static_cast<int>((range + sortOffset) * (static_cast<float>(rand()) / static_cast<float>(RAND_MAX)));
```

Veratil reviewed (commented):
The only other thing is perhaps it would be better to pull out all the `range + sortOffset` into a new variable so it's not repeated over and over and over and over. :)

zonkmachine commented:
@szeli1 There are some good points made in the original issue that you may want to read through. https://github.com/LMMS/lmms/issues/6499#issuecomment-2076103825

szeli1 commented:
I will not redo this now, but later. I also would like to add a feature that makes the instrument envelope apply to the arp notes like they were a whole note (using the base note's length).

zonkmachine commented:
> I will not redo this now, but later.

You mean the different orders of arpeggio mentioned in #6499? Merge now and refactor later?

szeli1 commented:
Yes. I do not want to merge yet. I can do the changes sooner if you would like that.

szeli1 commented:
Implemented changes, the vector can not be moved elsewhere since it can not be a member because of multithreading.
Using mutexes did not solve the issue, they just locked the threads for some reason.
Did not push jet because the above mentioned envelope feature.

szeli1 commented:
> Implemented changes, the vector can not be moved elsewhere since it can not be a member because of multithreading. Using mutexes did not solve the issue, they just locked the threads for some reason. Did not push jet because the above mentioned envelope feature.

szeli1 commented:
> I also would like to add a feature that makes the instrument envelope apply to the arp notes like they were a whole note (using the base note's length).

Can I add this to this pr?

zonkmachine commented:
> Can I add this to this pr?

If you want to do that change here I think we're fine with it. I usually fail to stick with just one change when I'm fiddling with the arpeggiator myself. ;)

szeli1 commented:
Order of the sorted notes in the arpeggio is now "octave-arp-piano". The implementation can be optimized by using binary search, but in my view that would be too complex without much performance improvements.

Added new check box in the envelope settings that makes the envelope use the parent note instead of the sub note.
This will allow the user to make the envelope apply over multiple arpeggio notes instead of each note separately.

Please check spelling.

zonkmachine commented:
This now seem to have issues. For instance. If I arpeggiate over multiple notes and one octave with sort, up is now down and the other way around. Please test this some more.

szeli1 commented:
Fixed the sorting algorithm and added more comments.
The arp settings should now behave like with the old sorting order.

zonkmachine commented:
> Added new check box in the envelope settings that makes the envelope use the parent note instead of the sub note. This will allow the user to make the envelope apply over multiple arpeggio notes instead of each note separately.


Interesting, I see what you're trying to do but I don't think this is ready for merge. Some quick observations.
* It's a bit hard to dial in an envelope that works both for the individual arpeggio notes and for the whole arpeggio, because they use the same envelope (this is my understanding anyway).
* I think I've seen similar implementations in synth plugins but they would let the arpeggio continue over the release.
* Also, in this test project you see that there is a bump in volume on key off where this is not the case with the new option disabled. 
[arpenvelope.mmp.txt](https://github.com/LMMS/lmms/files/15313310/arpenvelope.mmp.txt)

I think the envelope part is probably better to bud off to it's own PR.

szeli1 commented:
> envelope part is probably better to bud off to it's own PR.

I reverted the changes. Looking back on this, I completely agree with you.

zonkmachine commented:
I've tested it again but it doesn't seem to behave properly. The current implementation seem to be octave-piano-arp, referring to @DomClark from his analysis of the arpeggiator in #6499.

Borrowed from https://github.com/LMMS/lmms/issues/6499#issuecomment-2076103825
![image](https://github.com/LMMS/lmms/assets/1873225/474fc3d0-170d-4c28-b360-4690764edaa1)

Here are some different ways the notes (C-F-A) could be arpeggiated on.
1. The way LMMS does it today (piano-octave-arp)
2. How it currently works on this PR (octave-piano-arp)
3. The behavior I believe we agreed on (octave-arp-piano)
4. The original implementation in this PR (octave-sort(arp-piano))

Project used in testing: [arpsorttest.mmp.txt](https://github.com/user-attachments/files/15570100/arpsorttest.mmp.txt)

szeli1 commented:
Fixed, I really hope the arp is working as intended now.

zonkmachine commented on src/core/InstrumentFunctions.cpp:
```diff
@@ -435,26 +508,27 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
 		}
-		else if ((dir == ArpDirection::UpAndDown || dir == ArpDirection::DownAndUp) && range > 1)
+		else if ((dir == ArpDirection::UpAndDown || dir == ArpDirection::DownAndUp) &&
+			offsetRange > 1)
```
I'd keep this on the same line.

zonkmachine reviewed (approved):
Looks great and tests fine. I'm approving this and keep testing it ahead of the final word from @DomClark

DomClark reviewed (commented):
I think this PR can be simplified a bit. Firstly, I would introduce the variable `const auto arpMode = static_cast<ArpMode>(m_arpModeModel.value());`, to avoid repeating that expression every time you want to check the mode. Then, I would do the following:
* The `notMinIndex` bits can be replaced with an early return. I would check `arpType == ArpMode::Sort && _n != cnphv.first()` just before the declaration of `chord_table`, and return if it's true. I think this is a more obvious (and a bit more efficient) way to ensure only the first note drives the sorted arpeggio processing.
* Rather than introducing `offsetRange`, and replacing `range` with it, I would put the appropriate value into `range` in the first place. I'd rename the original `range` to `singleNoteRange` (or something like that), and then define `const auto range = arpType == ArpMode::Sort ? singleNoteRange * cnphv.size() : singleNoteRange;`. Then `total_range` can be removed too. This will simplify the diff a bit, and is a bit clearer as a variable name (I'm not sure what the "offset" refers to in "offsetRange").
* You can use a modulo operation to find which step of the chord you're at, in the same way you do to find which octave you're in. Then you don't need to build a list of notes in a separate vector, and can just sort `cnphv` in-place. The note computation will end up looking something like this:
  ```cpp
  const auto octaveDiv = std::div(cur_arp_idx, total_chord_size);
  const auto octave = octaveDiv.quot;
  const auto arpDiv = std::div(octaveDiv.rem, cnphv.size());
  const auto arpIndex = arpDiv.quot;
  const auto pianoRollIndex = arpDiv.rem;
  sub_note_key = cnphv[pianoRollIndex]->key()
  	+ chord_table.chords()[selected_arp][arpIndex]
  	+ octave * KeysPerOctave;
  ```
  See how the octave-arp-piano order becomes obvious now. The octave is the first quotient, the arp index the second, and the piano roll index the final remainder.

szeli1 commented:
> * You can use a modulo operation to find which step of the chord you're at

Smart idea, I didn't think about that.
Implemented the changes that you suggested.

DomClark reviewed (approved):
Code looks good to me, although I haven't tested it myself.

zonkmachine commented:
On the connection with #2606 (Race condition in arpeggio processing). Initially it looked fixed but this was just with the simple test file provided and I couldn't replicate the fix. That issue may have been more intermittent than I remember. I think this PR should be merged and #2606 be kept open.

zonkmachine commented:
@szeli1 Thank you for coding this, and for your patience in the review. Top notch work!
@DomClark Thanks for bringing clarity to the issue.
 @Veratil Thank you for reviewing.

This is a great step for the arpeggiator and for lmms-1.3
The arpeggiator now can be played via a keyboard live or played back from the piano
roll and it will work like any other 'arp' out there. With some extra cherries on top and
obviously a bug or two but the biggest glitches, in my opinion, have been taken care of.
